### ISSUE TITLE
qml ui

### ISSUE THREAD

**Issue by fuchsmich** [QTodoTxt/QTodoTxt#421]
Hi, I did a todo.txt app for SailfishOS https://github.com/fuchsmich/zutun.txt
It uses qml for the UI. Is there any interest in switching to qml for your project? (so we could share our code). If yes, I could start rebuilding the UI in qml.

**Comment by oroulet** [comment]
qml is the future and the UI code of qtodotxt does not follow the usual Qt conventions so I am not against a rewrite, but it will be quite a lot of work for you ;-)

**Comment by oroulet** [comment]
maybe start by rewrtting the TaskListView which displays tasks

**Comment by fuchsmich** [comment]
Ok. I guess the basics of the UI are rather simple in qml.
But I never did a python project, so i have no idea how to connect the 
python and qml part (data, signals, slots).
Anyways, a lot of people are using python, so its time for me to learn 
it. ;)

Am 28.03.2017 um 09:38 schrieb oroulet:

**Comment by oroulet** [comment]
much easier to learn python than C++. Anyway what you need to do is rather independent of the rest of the application. Make a qml widget with the methods that the TaskListView has: 
* createTask(task=None) (start editor)
* addTask(task)
* editTask(task)
* selectTask(task)
*removeTask(task)
That should be it

**Comment by fuchsmich** [comment]
SailfishOS supports pyotherside in its official app store ("harbour").
https://wiki.merproject.org/wiki/Sailfish/Python_Development#Python_and_Qt_5
Is this an option for you or would you like to stick to pyqt5?

**Comment by oroulet** [comment]
pyotherside seems to be a solution to replace javascript by python inside qml, isnt't? I would stick with js inside qml to avoid more dependencies

**Comment by fuchsmich** [comment]
ok, i see.
I guess the text in the link i posted is a bit missleading. I dont see 
any limitations in POS compared to pyqt.
But you have to touch c++ to use POS. With pyqt you can stay with python 
(+ qml (+ (some) js)).

Am 28.03.2017 um 11:01 schrieb oroulet:

**Comment by oroulet** [comment]
pyotherside seems to come with ubuntu. so no problem to use it if you think it helps. (I have very little experience with qml)

**Comment by oroulet** [comment]
As far as I understand we need pyqt5 anyway to bind python to qml. One could of course try to write a pure qml application using POS but then the entire application must be rewritten and it becomes an entire new project. No need to start from qtodotxt then.

What I think is possible is to replace progressively most pyqt5 code with qml. starting widget by widget. This would allow for a better looking application while not having to rewrite everything.

**Comment by topas-rec** [comment]
Does anyone know how qml is related to ubuntu convergence?

**Comment by topas-rec** [comment]
Yes i think this is supported.
I think Ubuntu touch (or convergence) needs a to-do app. ;-)

**Comment by fuchsmich** [comment]
You can use my sailfish app as template then. ;)

**Comment by fuchsmich** [comment]
Very quick hack:

https://github.com/fuchsmich/QTodoTxt/blob/qml/QTodoTxt.qml

![qtt](https://cloud.githubusercontent.com/assets/2899761/24415106/ab4681d2-13e0-11e7-8bf2-7cfe9e7404c0.png)

**Comment by topas-rec** [comment]
I thought about qtodotxt - which should also work - especially with qml.

**Comment by oroulet** [comment]
@fuchsmich looks OK although I still think we should start with widgets not the entire app, otherwise it will be too much work and a complete rewrite. For example TaskListView. It will interesting to see how we can handle a python object  (a Task) send to qml with a signal.... is it possible at all

**Comment by fuchsmich** [comment]
Yes. We can make it this way.
It was just a very quick hack to see, how I can launch a qml gui from 
python.
It works now. I forked the project and made a branch "qml".
In linux if you start bin/qtodotxt the gui will now show up.

The next step would be to export the tasklist from python to qml.

Maybe we could use some class derived from Listmodel (or similar model).
http://doc.qt.io/qt-5/qtquick-modelviewsdata-cppmodels.html


Am 29.03.2017 um 09:09 schrieb oroulet:

**Comment by oroulet** [comment]
inheriting from ListModel gives litlle value in python. You usually one maintain the internal model in a python list and populate the listmodel when necessary. In fact in qtodotx we use a qlistwidget, since I never found a simple way to display clickable links in a listview, while it is very easy with a listwidget by setting a label on it. you need to look at code og qtdodotxt anyway

**Comment by fuchsmich** [comment]
first success of data exchange between python and qml:

here: 
https://github.com/fuchsmich/QTodoTxt/blob/qml/qtodotxt/app.py#L155
the MainController instance gets exported as "mc" to qml.

here
https://github.com/fuchsmich/QTodoTxt/blob/qml/qtodotxt/ui/controllers/
main_controller.py#L58
I defined a pyqtProperty enabling the access to tasklist which is set in 
_applyFilters

same here
https://github.com/fuchsmich/QTodoTxt/blob/qml/qtodotxt/lib/tasklib.py#L96
so we can get the text

Juhuuu. :)
You can still use the original interface and see the changes of filtered tasks 
in the qml interface.

**Comment by oroulet** [comment]
Can you make a pull request? Just so I have a diff to look at?

**Comment by fuchsmich** [comment]
ok, I added some comments for you.

next step is propably to implement the details of the tasklistview in 
qml and - so one can interact with it - make the actions available to qml.

Am 02.04.2017 um 20:51 schrieb oroulet:
> Can you make a pull request? Just so I have a diff to look at?
>

**Comment by oroulet** [comment]
Btw I like your use of one task list as input to ui. That is how it should have been done. We could simply have two lists. One filtered and one not. Your work looks good, it will make it possible to do a huge cleanup

**Comment by fuchsmich** [comment]
I'm not sure, what you mean.

On Montag, 3. April 2017 18:18:38 CEST oroulet wrote:
> Btw I like your use of one task list as input to ui. That is how it should
> have been done. We could simply have two lists. One filtered and one not.
Are you referring to the sailfish app or this one?

> Your work looks good, it will make it possible to do a huge cleanup
Thanks. :)

**Comment by oroulet** [comment]
Btw . Use constant=True for your properties to avoid warnings.
And look at loaders and source properties to change delegate on events

**Comment by fuchsmich** [comment]
Hm, it seems like we have to redefine all actions in qml again because firstly 
they are not reuseable in qml and secondly they use qwidgets.

**Comment by oroulet** [comment]
that is not a big issue. make adding an task and editing one work then the rest will be easy. Will be some work to make the filter tree work too....
But I still think that you should NOT try to reimplement the entire ui in qml. Why not using https://doc.qt.io/qt-5/qquickview.html and only replace TaskListView in current UI with QML. When we have it working, we can work further on replacing the rest of the code.

**Comment by fuchsmich** [comment]
I c. Didn't know one can mix widgets and qml in one interface.

**Comment by oroulet** [comment]
Thinking about it maybe it is not bad to rewrite everything, .... But more work.... If you make adding task and editing one work, I can look at how to integrate the filter stuff

**Comment by fuchsmich** [comment]
I will think about it next days.
There is still a lot of work left making the tasklistview work.
When its finished, we can still decide how to continue.
(also see upcoming commit).

**Comment by oroulet** [comment]
one more thing that may be hard is to integrate text completion. We probably need to reuse the python inline editor inside qml (If you do not want to reimplement everything in qml, which is probably hard) . any idea how to do that?

**Comment by fuchsmich** [comment]
No, I ignored it successfully till now. ;)
But my hope is, I'm not the first who tries to implement this.


Am 05.04.2017 um 09:54 schrieb oroulet:

**Comment by oroulet** [comment]
I have an idea how it might work to use a pyqt widget inside qml. Should I try or are you also working on it? ( I can send a PR to your fork if it works)

**Comment by fuchsmich** [comment]
Yes, please try. Yesterday I just walked through the code around the 
completer.
I have some ideas for qml, but I think its better to use qcompleter if 
it integrates nicely.

Am 07.04.2017 um 09:06 schrieb oroulet:

**Comment by oroulet** [comment]
If you find a solution in pure qml, let me know. As far as I understood there is no qcompleter class so it is hard...

**Comment by oroulet** [comment]
Can you remove the start of the original UI, I do not think we need it? I will make a repository in qtodotxt we work on together

**Comment by fuchsmich** [comment]
I guess you just have to comment this line:

https://github.com/fuchsmich/QTodoTxt/blob/qml/qtodotxt/app.py#L168

Am 07.04.2017 um 12:26 schrieb oroulet:

**Comment by oroulet** [comment]
I had a look now. it looks like custom widget can only inherit from a kind of painter widget. One cannot use a higher level QWidget inside a qml application. But qml inside qwidget is possible. So it looks like the only possible solution here is too implement a completer in qml. you wrote that you had some ideas. Do you have a link or something?
Maybe I should focus on re-architecturing qtodotxt code to work with your qml gui. But we need to make sure we will find a solution for the completer stuff...

**Comment by oroulet** [comment]
https://wiki.qt.io/Qt_Quick_Best_Practices

**Comment by oroulet** [comment]
Feel free to fix improve that document: https://github.com/QTodoTxt/QTodoTxt/wiki/QTodotTxt2-QML-interface-architecture-ideas
So we get a good common understanding of what should be done

**Comment by oroulet** [comment]
created new repository. push directly to it without creating PR... as long as we do not create too much conflicts. But I am focusing on python code https://github.com/QTodoTxt/QTodoTxt2

**Comment by oroulet** [comment]
there is a core dump when leaving app. I think it was there before my changes... some qt objects without parents I guess... but no idea where..

**Comment by oroulet** [comment]
I cannot start app after your last update

/qtodotxt
QQmlApplicationEngine failed to load component
file:///home/olivier/dev/qml-qtodotxt/qml/QTodoTxt.qml:19 Type Preferences unavailable
file:///home/olivier/dev/qml-qtodotxt/qml/Preferences.qml:3 module "QtQuick.Controls" version 2.1 is not installed


I have qt5.7. is controls part of 5.8?

**Comment by fuchsmich** [comment]
I'm a bit confused which version to use in qtquick imports.
Downgraded to 1.4 in preferences.

**Comment by fuchsmich** [comment]
Yes, I also have this in Windows. In Linux it doesnt happen as far as i can 
see.

But I'm quite sure it will be gone, when everything is implemented neatly.

BTW I suggest the best way to use the maincontroller is to export the class to 
qml

qmlRegisterType(MainController, 'MC', 1, 0, 'MC')

import it there

import MC 1.0 MC

and instanciate the Class in qml

MC {
	id: mainController
}

**Comment by oroulet** [comment]
I do not know what the best way is. I let you decide. But I think it would be better not to use abbreviations like MC. use MainController or whatever easy to understand.

and move qml directory to qtodotxt folder. it will necessary anyway for installation

**Comment by fuchsmich** [comment]
Yes, of course and yes. :)

Am Sa. Apr. 8 20:32:11 2017 GMT+0200 schrieb oroulet:
> I do not know what the best way is. I let you decide. But I think it would be better not to use abbreviations like MC. use MainController or whatever easy to understand.
> 
> and move qml directory to qtodotxt folder. it will necessary anyway for installation
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/QTodoTxt/QTodoTxt/issues/421#issuecomment-29273688

-- 
Gesendet von meinem Jolla

**Comment by oroulet** [comment]
@fuchsmich keyboard can be used on filter tree but the mouse cannot select anything. Do you know how to fix that? thanks

**Comment by fuchsmich** [comment]
Hm, in Linux mouse works. Will check it tomorrow in Windows.

**Comment by oroulet** [comment]
I only use Linux. Ubuntu 17.04. if you also use Linux then we can just
ignore Windows for the time being

**Comment by oroulet** [comment]
The most work for the application. It is only missing for the treeview. Maybe I need to add a mouse area?

**Comment by fuchsmich** [comment]
My guess is, that it's a bug in your Qt version.

**Comment by oroulet** [comment]
@fuchsmich do a rebase with for example 'git pull -r' after local commit, so we avoid commits like this one  https://github.com/QTodoTxt/QTodoTxt2/commit/bb369cf8e92e862fd9e347a1497a3fb35b4cb8db

**Comment by oroulet** [comment]
and replace TextLine with TextArea in editor so we can do multiline edits. This was one of the common complaints of qtodotx. important to fix it from start ;-)

**Comment by fuchsmich** [comment]
Ok, thanks for the hint. It's my first project on github, which I do in 
cooperation.

Am 10.04.2017 um 09:34 schrieb oroulet:

**Comment by fuchsmich** [comment]
Ok. What is it for? Can one add multiple tasks at once?

I'm also thinking about an alternate implementation of the tasklist with 
tableview, as it implements some things we are using (e.g. line 
selection, highlighting, alterning row colors).


Am 10.04.2017 um 09:36 schrieb oroulet:

**Comment by oroulet** [comment]
Texarea is for editing long tasks. Tableview is fine

**Comment by oroulet** [comment]
I already committed change to texarea this morning. Tableview is fine, then
we can have a column for priority so people can order after them

**Comment by topas-rec** [comment]
> Tableview is fine, then we can have a column for priority so people can order after them 

Is that just an additional view? I think this won't work together with the inline editing. Am I wrong?
Btw i like the table view idea.

**Comment by oroulet** [comment]
Ideally, the columns would become one when editing... Not sure how to implement it..

**Comment by topas-rec** [comment]
Okay. What do you all think about dropdown fields for every column?
Text fields like task description have to stay just text. (Like a spreadsheet)
For the date the dropdown could show things like "tomorrow, today" and most important a calendar widget (if such a thing exists)
So a table view with cells to fill. Using small editors for every column. (Including simple dropdown things and text fields)
Sorting is nice. If the table view already has it - less work for us.

What do you think?

**Comment by fuchsmich** [comment]
I played around with tableview, but didn't like the results.
Listview is much more flexible and easier to layout.
Sorting can be easily implemented in python. Its just one additonal step after filtering the complete tasklist.
For quicksorting we could make a toolbar or buttonbar which acts like the column header of a tableview.

**Comment by fuchsmich** [comment]
TBH I don't like the idea of abstracting the textline too much (like putting each property of a task in its own cell). To me it's semantically too far away from todo.txt. And tableview wastes space: what about tasks having no priority, creationdate, duedate,... there would be empty cells.

If we do it then as an alternative, switchable view to listview.

I came up with tableview, because I thought things like highlighting or click events would be easier to handle. But to me listview is still better. I would have used tableview just with a single column so it looks like the listview now.

A calendar widget is a nice idea as its already available in QtQuick:
doc.qt.io/qt-5/qml-qtquick-controls-calendar.html

**Comment by oroulet** [comment]
my idea, a few month ago was to put the priority in its own column since anyway it must be at the beginning and we save place for it when generating html. then both columns would get merged in editor.  But I did not manage to make it work qith QWidget and I am not sure it is possible with qml either. so keep listview for now if it is easier

**Comment by fuchsmich** [comment]
if you make the priority a property of a task (like "text" and "html"), 
i can render it as you like and you can kick it out of the html-string.

Am 12.04.2017 um 15:04 schrieb oroulet:

**Comment by oroulet** [comment]
it is already a property, but let's look at it later.
Btw I have not looked at you completer code, but maybe I should export projects and contexts as properties from maincontroller so you can use them for completion?

**Comment by fuchsmich** [comment]
too late. :)

Yes, I need contexts and projects and all the other stuff which should 
be completed. Right now I'm using an array:
https://github.com/QTodoTxt/QTodoTxt2/blob/qml/qtodotxt/qml/TaskLine.qml#L91


Am 12.04.2017 um 15:19 schrieb oroulet:

**Comment by oroulet** [comment]
using an model is a godd idea!

**Comment by oroulet** [comment]
but modifying a task does not work anymore! you broke something in one of
the last commits...

**Comment by fuchsmich** [comment]
I see. The completion thing intercepts the return-key.

Am 12.04.2017 um 15:34 schrieb oroulet:
> but modifying a task does not work anymore! you broke something in one of
> the last commits...
> 
> On Wed, 12 Apr 2017 at 15:33 Olivier Roulet-Dubonnet <

**Comment by fuchsmich** [comment]
fixed.

Am 12.04.2017 um 15:34 schrieb oroulet:
> but modifying a task does not work anymore! you broke something in one of
> the last commits...
> 
> On Wed, 12 Apr 2017 at 15:33 Olivier Roulet-Dubonnet <
